include/swift/SIL/SILType.h

```diff
@@ -159,6 +159,17 @@ class SILType {
   }
 
   /// Returns the canonical AST type referenced by this SIL type.
+  ///
+  /// NOTE:
+  /// 1. The returned AST type may not be a proper formal type.
+  ///    For example, it may contain a SILFunctionType instead of a
+  ///    FunctionType.
+  /// 2. The returned type may not be the same as the original
+  ///    unlowered type that produced this SILType (even after
+  ///    canonicalization). If you need it, you must pass it separately.
+  ///    For example, `AnyObject.Type` may get lowered to
+  ///    `$@thick AnyObject.Type`, for which the AST type will be
+  ///    `@thick AnyObject.Type`.
```

Comment:
How about:

> Note that the AST type is not the same thing as the formal type that was lowered to produce this SIL type.  Swift uses largely the same AST classes for both the formal types used in the user-facing type system and the "lowered" representation types used by SIL.  Some type classes only appear in one or the other; for example, FunctionType is only used for formal types and is always a SILFunctionType in the SIL type system.  Other type classes have stronger requirements in the SIL type system; for example, MetatypeType must have a representation when it is used as a SIL value type.  Furthermore, and somewhat confusingly, SIL types may have formal types as components; for example, type arguments to generic types are usually formal types rather than lowered types (but Optional is a special case).  Regardless of these details, if you need the formal type, you must pass it down separately; there is no way to recover it given only a SILType.

Reply:
Couple of concerns: 

---

Afaict, the suggested comment, including this line

> Note that the AST type is not the same thing as the formal type that was lowered to produce this SIL type

does not imply what I wrote in point 1 earlier

> The returned AST type may not be a proper formal type.

For example, the reader might think, "oh, even if it doesn't return _the_ formal type that was used to produce this SIL type (clear from your comment or point 2 in mine), at least it returns _a_ well-formed formal type, right?" (this is literally the assumption I had before our discussion). However, that would be an incorrect assumption.

---

> Swift uses largely the same AST classes for both the formal types used in the user-facing type system and the "lowered" representation types used by SIL. Some type classes only appear in one or the other; for example, FunctionType is only used for formal types and is always a SILFunctionType in the SIL type system. Other type classes have stronger requirements in the SIL type system; for example, MetatypeType must have a representation when it is used as a SIL value type. Furthermore, and somewhat confusingly, SIL types may have formal types as components; for example, type arguments to generic types are usually formal types rather than lowered types (but Optional is a special case). Regardless of these details

I feel that `getASTType` is not the right place for this kind of high-level overview. If there's a need to say "regardless of the detail", it feels to me that the details should be someplace else, and that should be referenced here with a short comment like "Don't do X because it might fail in example Y, do X' instead. See Z for more information." where Z contains a high-level overview plus details.

If you can suggest a different place where this might fit better, I'm happy to add it to a different place. Adding it here seems a bit odd to me.

Reply:
You're right that it's probably best to just refer to the main SIL documentation on formal vs. lowered types and to make it clear that there's no way to recover a formal type in general from a lowered type.

Reply:
I've added an explicit reference to `docs/SIL.rst`.